refactor: make `MonoidAlgebra` into a one-field structure

YaelDillies opened the pull request:
---

- [x] depends on: #26732
- [x] depends on: #28407
- [x] depends on: #28460
- [x] depends on: #28511
- [x] depends on: #29207
- [x] depends on: #30877
- [x] depends on: #31143
- [x] depends on: #31144
- [x] depends on: #32253
- [x] depends on: #32254
- [x] depends on: #32398
- [x] depends on: #32501
- [x] depends on: #32508
- [x] depends on: #32562
- [x] depends on: #32588
- [x] depends on: #32591
- [x] depends on: #32592
- [x] depends on: #32604
- [x] depends on: #32965
- [x] depends on: #33093
- [x] depends on: #33094
- [x] depends on: #33099
- [x] depends on: #33137
- [x] depends on: #33139
- [x] depends on: #33141
- [x] depends on: #33452
- [x] depends on: #33453
- [x] depends on: #33482
- [x] depends on: #33492
- [x] depends on: #36675
- [ ] depends on: #36746

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -499,6 +499,9 @@ def of [MulOneClass G] : G →* MonoidAlgebra k G :=
     map_one' := rfl }
 
+@[simp]
+lemma linearCombination_of [MulOneClass G] : linearCombination k (of k G) = .id := by ext; simp
```
What's the type of this equality?

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

eric-wieser commented:
Is it a conscious choice not to do `AddMonoidAlgebra` at the same time?

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

JovanGerb commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -19,7 +22,7 @@ conditions at all. In this case the construction yields a not-necessarily-unital
 algebras to magmas, and we prove this as `MonoidAlgebra.liftMagma`.
 
-In this file we define `MonoidAlgebra k G := G →₀ k`, and `AddMonoidAlgebra k G`
+In this file we define `R[M] := G →₀ k`, and `AddMonoidAlgebra k G`
```
If we have the to_additive feature, do you think that the irreducible multiplication of AddMonoidAlgebra should still be defeq to the irreducible multiplication of MonoidAlgebra? Many proofs that rely on this defeq can instead be done by to_additive. But some theorems can inherently not be translated by to_additive, if one version uses `Multiplicative` and the other doesn't. And especially the (multiplicative) equivalence between AddMonoidAlgebra and MonoidAlgebra becomes harder to prove without this defeq.

YaelDillies replied:
My current hope is that we can make `AddMonoidAlgebra R M` be defeq to `MonoidAlgebra R (Multiplicative M)`.

JovanGerb replied:
Or define it as `MonoidAlgebra R M`, and rely on the defeq in `Multiplicative`?

YaelDillies replied:
> the (multiplicative) equivalence between AddMonoidAlgebra and MonoidAlgebra becomes harder to prove without this defeq.

I wouldn't worry too much about that. It's easy to do regardless

JovanGerb commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -87,712 +100,1077 @@ def unexpander : Lean.PrettyPrinter.Unexpander
+/-- The product of `x y : R[M]` is the finitely supported function whose value at `m` is the sum of
+`x m₁ * y m₂` over all pairs `m₁, m₂` such that `m₁ * m₂ = m`.
+(Think of the group ring of a group.) -/
+@[to_additive existing instMul]
+instance instMul : Mul R[M] where mul := mul'
```
The current definition of `MonoidAlgebra` has the following tags which are all very important:
```
@[to_additive (relevant_arg := 2) , to_additive_dont_translate]
```
Without `relevant_arg := 2`, you get what you describe: fixing R will fix `MonoidAlgebra`.

Without `to_additive_dont_translate`, it will translate `Mul (MonoidAlgebra R M)` into `Add (MonoidAlgebra)`.

So maybe try merging master and looking at the diff.

JovanGerb replied:
Ok, it turns out that there are a few more missing `(relevant_arg := 2)`, in this case for `addCommMonoid` and `smulZeroClass` (luckily these constants appear directly in the error message, so that you can figure out where the problem is)

JovanGerb replied:
The whole `relevant_arg := 2` thing could be avoided if we just put the (implicit) argument `M` as the first argument :)

YaelDillies replied:
Thanks! That fixed it, although swapping `R` and `M` really is a hack... Looking forward to having `relevant_arg` inferred automatically 😉

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -87,183 +93,56 @@ def unexpander : Lean.PrettyPrinter.Unexpander
+  coeffEquiv.nontrivial
+@[to_additive] instance unique [Subsingleton R] : Unique R[M] := coeffEquiv.unique
+@[to_additive] instance addCommMonoid : AddCommMonoid R[M] :=
+  coeffEquiv.addCommMonoid
```
```suggestion
  fast_instance% coeffEquiv.addCommMonoid
```
and for any other data-carrying instances you pull across

YaelDillies replied:
Are you sure your advice is fully accurate? If I use it on `inhabited` or `unique`, it complains with
```
Provided instance does not reduce to a constructor application
  coeffEquiv.unique
Reduces to an application of Equiv.unique.

Use `set_option trace.Elab.fast_instance true` to analyze the error.

Trace of fields visited: []Lean 4
```

eric-wieser replied:
`Equiv.unique` should be reducible, but isn't.

YaelDillies commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -465,63 +535,58 @@ section Monoid
 
 @[to_additive]
-instance semiring : Semiring (MonoidAlgebra R M) where
+instance semiring : Semiring R[M] where
```
See #30877

YaelDillies commented on Mathlib/Algebra/MonoidAlgebra/Opposite.lean:
```diff
@@ -25,36 +25,26 @@ namespace MonoidAlgebra
+-- /-- The opposite of a monoid algebra is equivalent as a ring to the opposite monoid algebra over the
+-- opposite ring. -/]
+@[simps! +simpRhs apply symm_apply]
 protected noncomputable def opRingEquiv : (MonoidAlgebra R M)ᵐᵒᵖ ≃+* MonoidAlgebra Rᵐᵒᵖ Mᵐᵒᵖ where
```
`attribute [to_additive self] MulOpposite.op_finsuppSum` says
```
`to_additive` validation failed: expected
  ∀ {ι : Type u_16} {M : Type u_17} {N : Type u_18} [inst : AddCommMonoid M] [inst_1 : Zero N] (f : ι →₀ N)
    (g : ι → N → M), AddOpposite.op (f.sum g) = f.sum fun i n ↦ AddOpposite.op (g i n)
but 'MulOpposite.op_finsuppSum' has type
  ∀ {ι : Type u_16} {M : Type u_17} {N : Type u_18} [inst : AddCommMonoid M] [inst_1 : Zero N] (f : ι →₀ N)
    (g : ι → N → M), op (f.sum g) = f.sum fun i n ↦ op (g i n)
```
😢

YaelDillies replied:
Ah, `attribute [to_additive self (dont_translate := M)] MulOpposite.op_finsuppSum` does it!

YaelDillies replied:
Do you think I should PR these `to_additive self` annotations, or will you fix `to_additive`'s heuristic instead?

JovanGerb replied:
If you leave a TODO in the code that is easy to search for, I can fix it later.

YaelDillies commented on Mathlib/Algebra/MonoidAlgebra/Defs.lean:
```diff
@@ -54,16 +52,28 @@ open Finsupp hiding single
 endowed with the convolution product. -/
-@[to_additive (relevant_arg := 2)
-/-- The additive monoid algebra over a semiring `R` generated by the additive monoid `M`.
+structure AddMonoidAlgebra (R M : Type*) [Semiring R] where
```
Let's make this future work 😇 There's already plenty to figure out here.

YaelDillies committed:
refactor: make `MonoidAlgebra` into a one-field structure

YaelDillies commented on Mathlib/Algebra/MonoidAlgebra/MapDomain.lean:
```diff
@@ -186,24 +224,21 @@ variable (M) in
-    ext
-    simp [mul_def]
-    simp [MonoidAlgebra, sum_mapRange_index, map_finsuppSum, single_apply, apply_ite]
+    classical ext; simp [mul_def, sum_mapRange_index, map_finsuppSum, single_apply, apply_ite]
```
What's your ETA on the fix to my problem being merged? a few weeks? a few months? In the first case, I'll wait. In the latter case, I'll work around it

YaelDillies committed:
broken

YaelDillies committed:
refactor(Algebra/MonoidAlgebra): use explicit conversion functions to/from `Finsupp`
Introduce `coeff : R[M] → M →₀ R` and `ofCoeff : (M →₀ R) → R[M]` as a way to convert explicitly between `MonoidAlgebra` and `Finsupp` and thus avoiding defeq abuse.

This is a fast and loose preliminary version of #25273. In particular, some proofs break and need to be fixed with either `set_option backward.isDefEq.respectTransparency false`, or by unfolding `MvPolynomial`. The situation will improve again once #25273 is in and further `MvPolynomial` is turned into an `abbrev`.

All the changes beyond introducing `coeff`/`ofCoeff` and using them in the basic API are motivated by fixing downstream breakage.

YaelDillies committed:
broken

YaelDillies committed:
chore: syntactically generalise `symm_mk`

YaelDillies committed:
chore(Algebra/MonoidAlgebra): make `uniqueRingEquiv_symm_apply` more useful
In essence, this changes the RHS from having support `univ` to having support `{1}`. These are equal, but having the RHS be `single 1 r` is much more useful in practice.

YaelDillies commented:
Reopened as #38714
